File: Example/ActionsListDemoObjectiveC/TabBarViewController.m
//
//  ActionsList
//  Copyright © 2018 LowKostKustomz. All rights reserved.
//

#import "TabBarViewController.h"
#import "ViewController.h"

@interface TabBarItemModel : NSObject

@property (nonatomic, strong, nullable) NSString* title;
@property (nonatomic, strong, nullable) UIImage* image;
@property (nonatomic, copy, nullable) void (^action)(TabBarItemModel*);
@property (nonatomic, strong, nullable) UIViewController* controller;
@property (nonatomic, strong, nullable) UIViewController* cachedViewController;

@end

@implementation TabBarItemModel

@end

@interface TabBarViewController ()

@end

@implementation TabBarViewController {
    TabBarItemModel* mainTabItem;
    TabBarItemModel* menuTabItem1;
    TabBarItemModel* menuTabItem2;
    TabBarItemModel* menuTabItem3;
    TabBarItemModel* menuTabItem4;
    
    NSMutableArray<TabBarItemModel*>* items;
    BOOL shouldReset;
    ActionsListModel* list;
}

- (instancetype)instantiate {
    TabBarViewController* controller = [[TabBarViewController alloc] init];
    return controller;
}

- (void)viewDidLoad {
    [super viewDidLoad];
    shouldReset = YES;
    
    [self setupItems];
    self.delegate = self;
}

- (void)viewWillAppear:(BOOL)animated {
    [super viewWillAppear:animated];
    
    if (shouldReset) {
        shouldReset = NO;
        [self resetItems];
    }
}

- (void) setupItems {
    mainTabItem = [[TabBarItemModel alloc] init];
    mainTabItem.title = @"Main";
    mainTabItem.image = [UIImage imageNamed:@"Home icon"];
    mainTabItem.controller = [[ViewController alloc] init];
    
    __weak TabBarViewController* weakSelf = self;
    
    menuTabItem1 = [[TabBarItemModel alloc] init];
    menuTabItem1.image = [UIImage imageNamed:@"List icon"];
    menuTabItem1.action = ^(TabBarItemModel* model) {
        [weakSelf showMenuForModel:model];
    };
    
    menuTabItem2 = [[TabBarItemModel alloc] init];
    menuTabItem2.title = @"Big Menu";
    menuTabItem2.image = [UIImage imageNamed:@"List icon"];
    menuTabItem2.action = ^(TabBarItemModel* model) {
        [weakSelf showMenuForModel:model];
    };
    
    menuTabItem3 = [[TabBarItemModel alloc] init];
    menuTabItem3.title = @"Bigger Menu";
    menuTabItem3.action = ^(TabBarItemModel* model) {
        [weakSelf showMenuForModel:model];
    };
    
    menuTabItem4 = [[TabBarItemModel alloc] init];
    menuTabItem4.title = @"Biggest Menu";
    menuTabItem4.image = [UIImage imageNamed:@"List icon"];
    menuTabItem4.action = ^(TabBarItemModel* model) {
        [weakSelf showMenuForModel:model];
    };
}

- (void)showMenuForModel:(TabBarItemModel*)model {
    if (model.controller) {
        if (model.controller.tabBarItem) {
            [self showMenuForItem:model.controller.tabBarItem model:model];
        }
    }
}

- (void)showMenuForItem:(UITabBarItem*)item model:(TabBarItemModel*)model {
    ActionsListModel* internalList = [item createActionsListWithDelegate:self];
    [internalList setupDismissAccessibilityHint:@"Double tap to dismiss"];
    NSString* title = model.title;
    
    if ([title isEqualToString:@"Big Menu"]) {
        [self addActions:4 toList:internalList];
    } else if ([title isEqualToString:@"Bigger Menu"]) {
        [self addActions:8 toList:internalList];
    } else if ([title isEqualToString:@"Biggest Menu"]) {
        [self addActions:16 toList:internalList];
    } else {
        [self addActions:2 toList:internalList];
    }
    
    [internalList presentWithCompletion:NULL];
    list = internalList;
}

- (void)addActions:(NSInteger)count toList:(ActionsListModel*)list {
    for (NSInteger i = 1; i <= count; i++) {
        NSMutableString* title = [@"" mutableCopy];
        
        for (NSInteger j = 0; j < i; j++) {
            [title appendString:[NSString stringWithFormat:@"Action number %ld", (long)i]];
            
            if (j != i - 1) {
                [title appendString:@"\n"];
            }
        }
        ActionsListDefaultButtonModel* actionModel = [[ActionsListDefaultButtonModel alloc] initWithLocalizedTitle:title
                                                                                                             image:[UIImage imageNamed:@"Dot"] action:^(ActionsListDefaultButtonModel* action) {
                                                                                                                 [action.list dismissWithCompletion:NULL];
                                                                                                             } isEnabled:YES];
        [actionModel setupAccessibilityWithAccessibilityLabel:[NSString stringWithFormat:@"%ld tests", (long)i]];
        [list addAction:actionModel];
    }
}

- (void)resetItems {
    items = [@[mainTabItem, menuTabItem1, menuTabItem2, menuTabItem3, menuTabItem4] mutableCopy];
    NSMutableArray<UIViewController*>* controllers = [@[] mutableCopy];
    for (NSInteger i = 0; i < items.count; i++) {
        TabBarItemModel* model = items[i];
        
        if (model.cachedViewController) {
            [controllers insertObject:model.cachedViewController atIndex:controllers.count];
            continue;
        }
        
        UIViewController* controller;
        if (model.controller) {
            controller = model.controller;
        } else {
            controller = [[UIViewController alloc] init];
        }
        controller = [[UINavigationController alloc] initWithRootViewController:controller];
        model.controller = controller;
        
        UITabBarItem* item = [[UITabBarItem alloc] initWithTitle:model.title image:model.image tag:i];
        if (!model.title) {
            CGFloat offset = 10;
            [item setImageInsets:UIEdgeInsetsMake(offset / 2, 0, -offset / 2, 0)];
        }
        
        item.titlePositionAdjustment = UIOffsetMake(0, -1);
        [controller setTabBarItem:item];
        
        [controllers insertObject:controller atIndex:[controllers count]];
    }
    [self setViewControllers:controllers];
}

- (void)actionsListDidHide {
    if (@available(iOS 11, *)) {
        [self.tabBar setTintColor:self.tabBar.unselectedItemTintColor];
    }
}

- (void)actionsListWillShow {
    [self.tabBar setTintColor:self.view.window.tintColor];
}

- (BOOL)tabBarController:(UITabBarController *)tabBarController shouldSelectViewController:(UIViewController *)viewController {
    NSUInteger index = [tabBarController.viewControllers indexOfObject:viewController];
    if (index && index < [items count]) {
        TabBarItemModel* item = [items objectAtIndex:index];
        if (item.cachedViewController) {
            [self setSelectedIndex:index];
            return true;
        } else if (item.action) {
            item.action(item);
            return false;
        }
    }
    return true;
}

@end

File: Example/ActionsListDemoSwift/TabBarViewController.swift
//
//  ActionsList
//  Copyright © 2017-2018 LowKostKustomz. All rights reserved.
//

import UIKit
import ActionsList

class TabBarItemModel {
    typealias ImageName = String
    
    enum TabBarItemType {
        case custom(String?, ImageName?)
    }
    
    enum TabBarItemActionType {
        case controller(UIViewController)
        case action((TabBarItemModel) -> ())
    }
    
    let type: TabBarItemType
    let actionType: TabBarItemActionType
    var viewController: UIViewController? = nil
    
    init(type: TabBarItemType,
         actionType: TabBarItemActionType) {
        self.type = type
        self.actionType = actionType
    }
}

class TabBarViewController: UITabBarController {
    
    private var mainTabItem: TabBarItemModel!
    private var menuTabItem1: TabBarItemModel!
    private var menuTabItem2: TabBarItemModel!
    private var menuTabItem3: TabBarItemModel!
    private var menuTabItem4: TabBarItemModel!
    
    fileprivate var items: [TabBarItemModel] = []
    private var shouldReset: Bool = true
    private var list: ActionsListModel?
    
    static func instantiate() -> TabBarViewController {
        let tabBar = TabBarViewController()
        return tabBar
    }

    override func viewDidLoad() {
        super.viewDidLoad()

        setupItems()
        delegate = self
    }
    
    override func viewWillAppear(_ animated: Bool) {
        super.viewWillAppear(animated)
        
        if shouldReset {
            shouldReset = false
            resetItems()
        }
    }
    
    private func setupItems() {
        mainTabItem = TabBarItemModel(type: .custom("Main", "Home icon"),
                                      actionType: .controller(ViewController()))
        
        menuTabItem1 = TabBarItemModel(type: .custom(nil, "List icon"),
                                       actionType: .action({ (model) in
                                        self.showMenu(fromModel: model)
                                       }))
        
        menuTabItem2 = TabBarItemModel(type: .custom("Big Menu", "List icon"),
                                       actionType: .action({ (model) in
                                        self.showMenu(fromModel: model)
                                       }))
        
        menuTabItem3 = TabBarItemModel(type: .custom("Bigger Menu", nil),
                                       actionType: .action({ (model) in
                                        self.showMenu(fromModel: model)
                                       }))
        
        menuTabItem4 = TabBarItemModel(type: .custom("Biggest Menu", "List icon"),
                                       actionType: .action({ (model) in
                                        self.showMenu(fromModel: model)
                                       }))
    }
    
    private func showMenu(fromModel model: TabBarItemModel) {
        if let controller = model.viewController,
            let item = controller.tabBarItem {
            showMenu(fromItem: item, model: model)
        }
    }
    
    private func showMenu(fromItem item: UITabBarItem,
                          model: TabBarItemModel) {
        guard let list = item.createActionsList(withDelegate: self)
            else {
                return
        }
        
        switch model.type {
        case .custom(let title, _):
            switch title ?? "" {
            case "Big Menu":
                addActions(4, toList: list)
            case "Bigger Menu":
                addActions(8, toList: list)
            case "Biggest Menu":
                addActions(16, toList: list)
            default:
                addActions(2, toList: list)
            }
        }
        
        list.present()
        self.list = list
    }
    
    private func addActions(_ count: Int,
                            toList list: ActionsListModel) {
        for i in 1...count {
            let title: String = [String](repeating: "Action number \(i)", count: i).joined(separator: "\n")
            
            list.add(action: ActionsListDefaultButtonModel(localizedTitle: title,
                                                           image: UIImage(named: "Dot"),
                                                           action: { action in
                                                            action.list?.dismiss()
            }))
        }
    }
    
    private func resetItems() {
        items = [mainTabItem, menuTabItem1, menuTabItem2, menuTabItem3, menuTabItem4]
        
        viewControllers = items.enumerated().map {
            let element = $0.element
            let index = $0.offset
            var viewController: UIViewController
            if let vc = element.viewController {
                viewController = vc
            } else {
                switch element.actionType {
                case .action:
                    viewController = UIViewController()
                case .controller(let controller):
                    viewController = controller
                }
                viewController = UINavigationController(rootViewController: viewController)
                element.viewController = viewController
            }
            
            var tabBarItem: UITabBarItem
            
            switch element.type {
            case .custom(let title, let imageName):
                tabBarItem = UITabBarItem(title: title, image: UIImage(named: imageName ?? ""), tag: index)
                if title == nil {
                    let offset: CGFloat = 10
                    tabBarItem.imageInsets = UIEdgeInsets(top: offset / 2, left: 0, bottom: -offset / 2, right: 0)
                }
            }
            
            tabBarItem.titlePositionAdjustment = UIOffset(horizontal: 0, vertical: -1)
            viewController.tabBarItem = tabBarItem
            return viewController
        }
    }
}

extension TabBarViewController: ActionsListDelegate {
    func actionsListWillShow() {
        if #available(iOS 10, *) {
            tabBar.tintColor = tabBar.unselectedItemTintColor
        }
    }
    
    func actionsListDidHide() {
        tabBar.tintColor = view.window?.tintColor
    }
}

extension TabBarViewController: UITabBarControllerDelegate {
    func tabBarController(_ tabBarController: UITabBarController, shouldSelect viewController: UIViewController) -> Bool {
        if let index = tabBarController.viewControllers?.index(of: viewController),
            index < items.count {
            let item = items[index]
            switch item.actionType {
            case .action(let action):
                action(item)
                return false
            case .controller:
                self.selectedIndex = index
                return true
            }
        }
        return true
    }
}
